File: feature-tests/test/com/github/joerodriguez/sbng2ex/LoginTest.java
package com.github.joerodriguez.sbng2ex;

import org.fluentlenium.adapter.FluentTest;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.test.SpringApplicationConfiguration;
import org.springframework.boot.test.WebIntegrationTest;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
import org.springframework.transaction.annotation.Transactional;

import java.util.concurrent.TimeUnit;

@RunWith(SpringJUnit4ClassRunner.class)
@SpringApplicationConfiguration(ExampleApiApplication.class)
@WebIntegrationTest("server.port:0")
public class LoginTest extends FluentTest {

    @Value("${local.server.port}")
    int port;

    @Autowired
    PasswordEncoder passwordEncoder;

    @Autowired
    JdbcTemplate jdbcTemplate;

    @Before
    public void setUp() {
        DbTruncatorKt.DbTruncator(jdbcTemplate);

        jdbcTemplate.update(
                "INSERT INTO users (email, password) VALUES (?,?)",
                new Object[]{"tester@example.com", passwordEncoder.encode("tester-good-password")}
        );
    }

    @Test
    public void testLogin_withBadEmail() {
        goTo("http://localhost:" + port + "/");
        await().atMost(5, TimeUnit.SECONDS).until("body").containsText("Login");

        fill("#email").with("tester-unknown@example.com");
        fill("#password").with("tester-good-password");

        findFirst("input[type=submit]").click();

        await().atMost(5, TimeUnit.SECONDS).until("body").containsText("Login failed");
    }

    @Test
    public void testLogin_withBadPassword() {
        goTo("http://localhost:" + port + "/");
        await().atMost(5, TimeUnit.SECONDS).until("body").containsText("Login");

        fill("#email").with("tester@example.com");
        fill("#password").with("tester-bad-password");

        findFirst("input[type=submit]").click();

        await().atMost(5, TimeUnit.SECONDS).until("body").containsText("Login failed");
    }

    @Test
    public void testLogin() {
        goTo("http://localhost:" + port + "/");
        await().atMost(5, TimeUnit.SECONDS).until("body").containsText("Login");

        fill("#email").with("tester@example.com");
        fill("#password").with("tester-good-password");

        click("input[type=submit]");

        await().atMost(5, TimeUnit.SECONDS).until("body").containsText("Login succeeded");
    }
}

File: feature-tests/test/com/github/joerodriguez/sbng2ex/feature/LoginTest.kt
package com.github.joerodriguez.sbng2ex.feature

import com.github.joerodriguez.sbng2ex.ExampleApiApplication
import com.github.joerodriguez.sbng2ex.dbtestinglibrary.truncateDb
import io.damo.kspec.spring.SpringSpec
import net.codestory.simplelenium.FluentTest
import org.springframework.boot.test.SpringApplicationConfiguration
import org.springframework.boot.test.WebIntegrationTest
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.test.context.ActiveProfiles

@SpringApplicationConfiguration(ExampleApiApplication::class)
@WebIntegrationTest("server.port:0")
@ActiveProfiles("test")
class LoginTest : SpringSpec({

    val port = injectValue("local.server.port", Int::class)
    val passwordEncoder = inject(PasswordEncoder::class)
    val fluentTest = FluentTest()

    val jdbcTemplate = inject(JdbcTemplate::class)

    before {
        jdbcTemplate.truncateDb()
        jdbcTemplate.update(
                "INSERT INTO users (email, password) VALUES (?,?)",
                *arrayOf<Any>("tester@example.com", passwordEncoder.encode("tester-good-password")))
    }

    describe("#login") {
        test("when email is bad") {
            val form = fluentTest.goTo("http://localhost:$port/").find("#login")

            form.find("#email").fill("tester-unknown@example.com")
            form.find("#password").fill("tester-good-password")

            form.find("input[type=submit]").first().click()
            form.should().contain("Login failed")
        }

        test("when password is bad") {
            val form = fluentTest.goTo("http://localhost:$port/").find("#login")

            form.find("#email").fill("tester@example.com")
            form.find("#password").fill("tester-bad-password")

            form.find("input[type=submit]").first().click()
            form.should().contain("Login failed")
        }

        test("when successful") {
            val form = fluentTest.goTo("http://localhost:$port/").find("#login")

            form.find("#email").fill("tester@example.com")
            form.find("#password").fill("tester-good-password")

            form.find("input[type=submit]").first().click()
            form.should().contain("Login succeeded")
        }
    }
})
